pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/IdenticalPredicateFilterOptimizer.java

```diff
@@ -0,0 +1,108 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.optimizer.filter;
+
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.request.Expression;
+import org.apache.pinot.common.request.Function;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.sql.FilterKind;
+
+
+/**
+ * This optimizer converts all predicates where the left hand side == right hand side to
+ * a simple TRUE/FALSE literal value. While filters like, WHERE 1=1 OR "col1"="col1" are not
+ * typical, they end up expensive in Pinot because they are rewritten as A-A==0.
+ */
+public class IdenticalPredicateFilterOptimizer extends BaseAndOrBooleanFilterOptimizer {
+
+  @Override
+  boolean canBeOptimized(Expression filterExpression, @Nullable Schema schema) {
+    // if there's no function call, there's no lhs or rhs
+    return filterExpression.getFunctionCall() != null;
+  }
+
+  @Override
+  Expression optimizeChild(Expression filterExpression, @Nullable Schema schema) {
+    Function function = filterExpression.getFunctionCall();
+    FilterKind kind = FilterKind.valueOf(function.getOperator());
+    switch (kind) {
+      case EQUALS:
+        if (hasIdenticalLhsAndRhs(filterExpression)) {
+          setExpressionToBoolean(filterExpression, true);
+        }
+        break;
+      case NOT_EQUALS:
+        if (hasIdenticalLhsAndRhs(filterExpression)) {
+          setExpressionToBoolean(filterExpression, false);
+        }
+        break;
+      default:
+        break;
+    }
+    return filterExpression;
+  }
+
+  /**
+   * Pinot queries of the WHERE 1 != 1 AND "col1" = "col2" variety are rewritten as
+   * 1-1 != 0 AND "col1"-"col2" = 0. Therefore, we check specifically for the case where
+   * the operand is set up in this fashion.
+   *
+   * We return false specifically after every check to ensure we're only continuing when
+   * the input looks as expected. Otherwise, it's easy to for one of the operand functions
+   * to return null and fail the query.
+   */
+  private boolean hasIdenticalLhsAndRhs(Expression operand) {
+    Function function = operand.getFunctionCall();
+    if (function == null) {
+      return false;
+    }
+    List<Expression> children = function.getOperands();
+    boolean hasTwoChildren = children.size() == 2;
```

Comment:
(minor) Is there any case `EQ` or `NEQ` can have other than 2 children? We can probably make a precondition

Reply:
a `precondition` would fail the query, no? even if it is possible, this function is really only supposed to optimize this one case.